components/canvas/canvas_data.rs

```diff
@@ -58,10 +57,118 @@ impl PathState {
+        &self,
+        control_point1: &Point2D<f32>,
+        control_point2: &Point2D<f32>,
+        control_point3: &Point2D<f32>,
+    );
+    fn close(&self);
+    fn ellipse(
+        &self,
+        origin: Point2D<f32>,
+        radius_x: f32,
+        radius_y: f32,
+        rotation_angle: f32,
+        start_angle: f32,
+        end_angle: f32,
+        anticlockwise: bool,
+    );
+    fn get_current_point(&self) -> Point2D<f32>;
+    fn line_to(&self, point: Point2D<f32>);
+    fn move_to(&self, point: Point2D<f32>);
+    fn quadratic_curve_to(&self, control_point: &Point2D<f32>, end_point: &Point2D<f32>);
+    fn finish(&self) -> Path;
+}
+
+impl GenericPathBuilder for azure_hl::PathBuilder {
+    fn arc(
+        &self,
+        origin: Point2D<f32>,
+        radius: f32,
+        start_angle: f32,
+        end_angle: f32,
+        anticlockwise: bool,
+    ) {
+        self.arc(
+            origin as Point2D<AzFloat>,
+            radius as AzFloat,
+            start_angle as AzFloat,
+            end_angle as AzFloat,
+            anticlockwise,
+        );
+    }
+    fn bezier_curve_to(
+        &self,
+        control_point1: &Point2D<f32>,
+        control_point2: &Point2D<f32>,
+        control_point3: &Point2D<f32>,
+    ) {
+        self.bezier_curve_to(
+            control_point1 as &Point2D<AzFloat>,
+            control_point2 as &Point2D<AzFloat>,
+            control_point3 as &Point2D<AzFloat>,
+        );
+    }
+    fn close(&self) {
+        self.close();
+    }
+    fn ellipse(
+        &self,
+        origin: Point2D<f32>,
+        radius_x: f32,
+        radius_y: f32,
+        rotation_angle: f32,
+        start_angle: f32,
+        end_angle: f32,
+        anticlockwise: bool,
+    ) {
+        self.ellipse(
+            origin as Point2D<AzFloat>,
+            radius_x as AzFloat,
+            radius_y as AzFloat,
+            rotation_angle as AzFloat,
+            start_angle as AzFloat,
+            end_angle as AzFloat,
+            anticlockwise,
+        );
+    }
+    fn get_current_point(&self) -> Point2D<f32> {
+        let AzPoint { x, y } = self.get_current_point();
+        Point2D::new(x as f32, y as f32)
+    }
+    fn line_to(&self, point: Point2D<f32>) {
+        self.line_to(point as Point2D<AzFloat>);
+    }
+    fn move_to(&self, point: Point2D<f32>) {
+        self.move_to(point as Point2D<AzFloat>);
+    }
+    fn quadratic_curve_to(&self, control_point: &Point2D<f32>, end_point: &Point2D<f32>) {
+        self.quadratic_curve_to(
+            control_point as &Point2D<AzFloat>,
+            end_point as &Point2D<AzFloat>,
+        );
+    }
+    fn finish(&self) -> Path {
+        Path::Azure(self.finish())
+    }
+}
+
 /// A wrapper around a stored PathBuilder and an optional transformation that should be
 /// applied to any points to ensure they are in the matching device space.
 struct PathBuilderRef<'a> {
-    builder: &'a PathBuilder,
+    builder: &'a Box<GenericPathBuilder>,
```

Comment:
This can use `&'a GenericPathBuilder` instead.

Reply:
(this still needs fixing)